### ISSUE TITLE
Bug

### ISSUE THREAD

**Issue by lucilfer12** [near-daos/sputnik-dao-contract#241]
# sputnik-staking — Inverted Guard in `storage_unregister` Permanently Burns the User's Staked Vote Tokens

**Target:** `near-daos/sputnik-dao-contract`, package `sputnik-staking`, file `sputnik-staking/src/storage_impl.rs`, function `storage_unregister` (NEP-145 `StorageManagement` implementation)
**HackenProof category hit:** Stealing or loss of funds

## Verification status

Code-confirmed. This is a self-contained logic defect verifiable purely from `storage_impl.rs` and `user.rs`'s field definitions — it does not depend on any external protocol nuance, so this report carries the same confidence as a fully-traced code path even without a compiled/executed PoC (no `rustc`/network available in this environment).

## Root cause

```rust
// sputnik-staking/src/storage_impl.rs
#[allow(unused_variables)]
#[payable]
fn storage_unregister(&mut self, force: Option<bool>) -> bool {
    assert_one_yocto();
    let account_id = env::predecessor_account_id();
    if let Some(user) = self.internal_get_user_opt(&account_id) {
        // TODO: figure out force option logic.
        assert!(user.vote_amount.0 > 0, "ERR_STORAGE_UNREGISTER_NOT_EMPTY");
        self.users.remove(&account_id);
        Promise::new(account_id).transfer(user.near_amount).detach();
        true
    } else {
        false
    }
}
```

Two distinct fields exist on `User` (`sputnik-staking/src/user.rs`): `near_amount` (NEAR the user paid purely to cover their storage row) and `vote_amount` (the actual balance of the fungible vote/governance token the user deposited into this contract via `ft_on_transfer`, custodied by the contract). These are tracked, and released, independently elsewhere in the contract — `internal_withdraw` reduces `vote_amount` and triggers a real `ft_transfer` back to the user; `storage_withdraw` (a separate NEP-145 method) releases only the spare `near_amount`.

`storage_unregister` conflates them incorrectly:

1. **The guard is inverted.** The NEP-145 standard's usual contract for `storage_unregister` is: refuse to unregister an account that still holds a balance, unless `force=true`. Here, the condition is backwards — it requires `vote_amount.0 > 0` to *proceed*, and panics with `"...NOT_EMPTY"` when the account **is** empty. The `force: Option<bool>` parameter required by the standard interface is accepted but marked `#[allow(unused_variables)]` and never read.
2. **The `force`-equivalent path never returns the underlying tokens.** When the (inverted) guard does pass — i.e., precisely when the user **still holds** a nonzero `vote_amount` — the function deletes the entire `User` record via `self.users.remove(&account_id)` and refunds only `near_amount`. No `ft_transfer` is issued for `vote_amount`, and `self.total_amount` (the contract-wide running total, returned by `ft_total_supply()`) is never decremented to reflect the removal.

The `// TODO: figure out force option logic.` comment directly on this code confirms it was left intentionally unfinished rather than designed this way.

## Impact

For any account that calls `storage_unregister()` while holding a nonzero `vote_amount` (the only state in which this inverted guard lets the call proceed at all):

- **The user's entire deposited vote-token balance becomes permanently unreachable.** The contract retains custody of the real fungible tokens (via the external `vote_token_id` FT contract), but the only bookkeeping row that could ever authorize their withdrawal is deleted in the same call. No other method in this contract can recover them for this account.
- **`self.total_amount` becomes permanently overstated** by the abandoned amount, corrupting `ft_total_supply()` for every future caller/integration that relies on it as an accurate figure.
- **If any portion of `vote_amount` was currently `delegate()`d to other accounts**, that delegated weight is never withdrawn from the DAO's own `total_delegation_amount` governance bookkeeping (`sputnikdao2/delegation.rs`) — no `undelegate` call is triggered here — leaving a permanent, un-revocable "phantom" contribution to another account's voting power.

## Who can trigger this, and how

This requires the account holder's own `storage_unregister()` call (1 yoctoNEAR attached, per `assert_one_yocto()`) while `vote_amount > 0` — it is not a bug an attacker can trigger against an unwilling victim directly. It is nonetheless a real fund-loss vector because `storage_unregister` is a **standardized NEP-145 method name**: wallets, dashboards, and generic "clean up my registered contracts" tooling call it by convention across many different contracts without necessarily knowing this specific contract inverts the usual guard. A tool built against the standard's normal assumption (block when non-empty, proceed when empty) is precisely the kind of caller most likely to hit this contract's non-standard branch by surprise while a user still has an active stake.

## Why this isn't "intended design"

The inverted condition contradicts both the NEP-145 standard's documented behavior and the plain meaning of the panic message itself (`"...NOT_EMPTY"` firing when the account **is** empty is a direct textual contradiction). Combined with the explicit `TODO` comment, this reads as unfinished logic rather than a deliberate choice — no design rationale would explain silently discarding a user's tokens and desyncing the total supply as the intended behavior of a standard interface method.

## Suggested fix direction

Implement the standard's actual semantics: block unregistration while `vote_amount > 0` unless `force == Some(true)`; and when forcing removal with a nonzero balance, either (preferred) issue an `ft_transfer` of the outstanding `vote_amount` back to the user before removing their record and decrement `self.total_amount` accordingly, or, if forfeiting is truly intended for the force path, at minimum keep `self.total_amount` correct and document the forfeiture explicitly.

## Severity note

Impact is a real, uncapped (up to the caller's full deposited balance) and irreversible loss of funds, which is explicitly in-scope under "Stealing or loss of funds." It is self-triggered rather than attacker-forced onto a victim, which should inform severity relative to a pure privilege-escalation bug — final classification rests with HackenProof triage.
